sequencing-server/src/utils/hasura.ts

```diff
@@ -0,0 +1,101 @@
+ * @param authSessionVariables If the request has gone through a hasura action this object will be populated.
+ * @param authHeader If the user sends a POST request directly to the endpoint with a bearer token this will be populated.
+ * @returns The AuthSessionVeriables of the user sending the request.
+ */
+export function getHasuraSession(
+  authSessionVariables: AuthSessionVariables | undefined,
+  authHeader: string | undefined,
+): AuthSessionVariables {
+  if (authSessionVariables !== undefined) {
+    return authSessionVariables;
+  } else if (authHeader !== undefined) {
+    return (jwtDecode(authHeader) as Jwt)['https://hasura.io/jwt/claims'];
+  }
+
+  throw new Error('Could not determine the user sending the request');
+}
+
+/**
+ * A helper function to get the username of the user who sent the request.
+ *
+ * @param authSessionVariables If the request has gone through a hasura action this object will be populated.
+ * @param authHeader If the user sends a POST request directly to the endpoint with a bearer token this will be populated.
+ * @returns The username of the user sending the request.
+ */
+export function getUsername(
+  authSessionVariables: AuthSessionVariables | undefined,
+  authHeader: string | undefined,
+): string {
+  if (authSessionVariables !== undefined) {
+    return authSessionVariables['x-hasura-user-id'];
+  } else if (authHeader !== undefined) {
+    return (jwtDecode(authHeader) as Jwt).username;
+  }
+
+  throw new Error('Could not determine the user sending the request');
+}
+
+/**
+ * This function checks to see if request coming from the Hasura action is being executed
+ * by a user with enough permissions. To check this we need to call a GQL query with the
+ * action key and the user session. If we ever merge the databases this should be changed
+ * to call the db function that does the permission checking directly.
+ *
+ * Some endpoints don't need to be checked which is why we have the mappings of endpoints
+ * to actions to check.
+ *
+ * @param url The endpoint that we're checking the permissions for.
+ * @param graphqlClient The GQL Client we're using to make the permission check.
+ * @returns True if the user is able to call the given action, false otherwise.
+ */
+export async function canUserPerformAction(
+  url: string,
+  graphqlClient: GraphQLClient,
+  hasuraSession: AuthSessionVariables,
+): Promise<boolean> {
+  if (url in ENDPOINTS_TO_ACTION_KEY) {
+    const permissionCheckQuery = await graphqlClient.request(
+      gql`
+        query checkUserPermissions($role: user_roles_enum!, $key: String!) {
+          user_role_permission_by_pk(role: $role) {
+            action_permissions(path: $key)
+          }
+        }
+      `,
+      {
+        role: hasuraSession['x-hasura-role'],
+        key: ENDPOINTS_TO_ACTION_KEY[url],
+      },
+    );
+
+    if (permissionCheckQuery.user_role_permission_by_pk.action_permissions === null) {
+      return false;
+    }
+  }
+
+  return true;
```

Comment:
I highly recommend inverting this logic to deny by default/fall-through rather than allow by default/fall-through. Harder to accidentally allow unintentional access that way.

Reply:
The problem is we have some endpoints that are unprotected, I could be exhaustive in the mapping but currently `health`, `get-command-typescript`, and `get-activity-typescript` don't do any permission checking.

Reply:
You could rewrite the functions as:
```ts
if (!url in ENDPOINTS_TO_ACTION_KEY) return true;
// The rest of the logic
```

Reply:
Could resolve that with an explicit whitelist of unprotected endpoints.

I'm just considering the case where another endpoint gets added and the map doesn't get updated, accidentally exposing endpoints that should be protected.

Reply:
I'm going to add a whitelist set, I think that's more readable and then we won't automatically allow usage on new endpoints.